src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,169 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> first;
+    private Node<T> last;
+    private int size;
+    private Node<T> boofNode;
+    private Node<T> newNode;
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            linkFirst(value);
+        } else {
+            linkLast(value);
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == 0) {
+            linkFirst(value);
+            return;
+        }
+        if (index == size) {
+            linkLast(value);
+            return;
+        }
+        isIndexOutOfBounds(index);
+        Node<T> oldNode = findeIndex(index);
+        newNode = new Node<>(oldNode.prev, value, oldNode);
+        oldNode.prev.next = newNode;
+        oldNode.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T newNode : list) {
+            linkLast(newNode);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isIndexOutOfBounds(index);
+        return findeIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        isIndexOutOfBounds(index);
+        boofNode = findeIndex(index);
+        T oldValue = boofNode.value;
+        boofNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        isIndexOutOfBounds(index);
+        boofNode = findeIndex(index);
+        T removedValue = boofNode.value;
+        if (index == 0) {
+            first = first.next;
+        } else if (index == size - 1) {
+            last = last.prev;
+        } else {
+            boofNode.prev.next = boofNode.next;
+            boofNode.next.prev = boofNode.prev;
+        }
+        size--;
+        return removedValue;
     }
 
     @Override
     public boolean remove(T object) {
+        for (int i = 0; i < size; i++) {
+            boofNode = findeIndex(i);
+            if (object == boofNode.value || object != null && object.equals(boofNode.value)) {
+                remove(i);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> findeIndex(int index) {
+        int halfSize = size / 2;
+        if (index < halfSize) {
+            boofNode = first;
+            if (index == 0) {
+                return boofNode;
+            }
+            int count = 1;
+            while (count <= index) {
+                boofNode = boofNode.next;
+                count++;
+            }
```

Comment:
this if check is redundant
and instead of while loop with counter let's use for i = 0 ...